README.md
# Fluent Plugin for Amazon Kinesis

 [![Build Status](https://travis-ci.org/awslabs/aws-fluent-plugin-kinesis.svg?branch=master)](https://travis-ci.org/awslabs/aws-fluent-plugin-kinesis)

## Overview

[Fluentd](http://fluentd.org/) output plugin
that sends events to [Amazon Kinesis](https://aws.amazon.com/kinesis/).

## Installation

This fluentd plugin is available as the `fluent-plugin-kinesis` gem from RubyGems.

    gem install fluent-plugin-kinesis

Or you can install this plugin for [td-agent](https://github.com/treasure-data/td-agent) as:

    fluent-gem install fluent-plugin-kinesis

If you would like to build by yourself and install, please see the section below.
Your need [bundler](http://bundler.io/) for this.

In case of using with Fluentd:
Fluentd will be also installed via the process below.

    git clone https://github.com/awslabs/aws-fluent-plugin-kinesis.git
    cd aws-fluent-plugin-kinesis
    bundle install
    rake build
    rake install

Also, you can use this plugin with td-agent:
You have to install td-agent before installing this plugin.

    git clone https://github.com/awslabs/aws-fluent-plugin-kinesis.git
    cd aws-fluent-plugin-kinesis
    bundle install
    rake build
    fluent-gem install pkg/fluent-plugin-kinesis

Or just download specify your Ruby library path.
Below is the sample for specifying your library path via RUBYLIB.

    git clone https://github.com/awslabs/aws-fluent-plugin-kinesis.git
    cd aws-fluent-plugin-kinesis
    bundle install
    export RUBYLIB=$RUBYLIB:/path/to/aws-fluent-plugin-kinesis/lib

## Dependencies

 * Ruby 1.9.3+
 * Fluentd 0.10.43+

## Basic Usage

Here are general procedures for using this plugin:

 1. Install.
 1. Edit configuration
 1. Run Fluentd or td-agent

You can run this plugin with Fluentd as follows:

 1. Install.
 1. Edit configuration file and save it as 'fluentd.conf'.
 1. Then, run `fluentd -c /path/to/fluentd.conf`

To run with td-agent, it would be as follows:

 1. Install.
 1. Edit configuration file provided by td-agent.
 1. Then, run or restart td-agent.

## Configuration

Here are items for Fluentd configuration file.

To put records into Amazon Kinesis,
you need to provide AWS security credentials.
If you provide aws_key_id and aws_sec_key in configuration file as below,
we use it. You can also provide credentials via environment variables as
AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY.  Also we support IAM Role for
authentication. Please find the [AWS SDK for Ruby Developer Guide](http://docs.aws.amazon.com/AWSSdkDocsRuby/latest/DeveloperGuide/ruby-dg-setup.html)
for more information about authentication.
We support all options which AWS SDK for Ruby supports.

### type

Use the word 'kinesis'.

### stream_name

Name of the stream to put data.

### aws_key_id

AWS access key id.

### aws_sec_key

AWS secret key.

### region

AWS region of your stream.
It should be in form like "us-east-1", "us-west-2".
Refer to [Regions and Endpoints in AWS General Reference](http://docs.aws.amazon.com/general/latest/gr/rande.html#ak_region)
for supported regions.

### http_proxy

Proxy server, if any.
It should be in form like "http://squid:3128/"

### endpoint

Default endpoint is aws. But you can specify the endpoint directly.
For example, you can connect to local endpoint, such as [Kinesalite](https://github.com/mhart/kinesalite).

### random_partition_key

Boolean. If true, the plugin uses randomly generated
partition key for each record. Note that this parameter
overrides *partition_key*, *partition_key_expr*,
*explicit_hash_key* and *explicit_hash_key_expr*.

### partition_key

A key to extract partition key from JSON object.

### partition_key_expr

A Ruby expression to extract partition key from JSON object.
We treat your expression as below.

    a_proc = eval(sprintf('proc {|record| %s }', YOUR_EXPRESSION))
    a_proc.call(record)

You should write your Ruby expression that receives input data
as a variable 'record', process it and return it. The returned
value will be used as a partition key. For use case example,
see 'Configuration examples' part.

### explicit_hash_key

A key to extract explicit hash key from JSON object.
Explicit hash key is hash value used to explicitly
determine the shard the data record is assigned to
by overriding the partition key hash.

### explicit_hash_key_expr

A Ruby expression to extract explicit hash key from JSON object.
Your expression will be treat in the same way as we treat partition_key_expr.

### order_events

Boolean. By enabling it, you can strictly order events in Amazon Kinesis,
according to arrival of events. Without this, events will be coarsely ordered
based on arrival time. For detail,
see [Using the Amazon Kinesis Service API](http://docs.aws.amazon.com/kinesis/latest/dev/kinesis-using-api-java.html#kinesis-using-api-defn-sequence-number).

### multi_records_per_put
Boolean. Default is false. By enabling it, you can group multiple records by partition_key. While when *random_partition_key* is enabled, all the records will be gathered together into one record, not grouped by partition_key . This feature is used to save money by descrising the number of kinesis put.

### multi_records_separator
String. Default is "\v". Yan can change by
```
multi_records_separator "#{11.chr}"
```
11 is the ASCII code of "\v"

Please note that if you set *detach_process* or *num_threads greater than 1*,
this option will be ignored.

### detach_process

Integer. Optional. This defines the number of parallel processes to start.
This can be used to increase throughput by allowing multiple processes to
execute the plugin at once. This cannot be used together with **order_events**.
Setting this option to > 0 will cause the plugin to run in a separate
process. The default is 0.

### num_threads

Integer. The number of threads to flush the buffer. This plugin is based on
Fluentd::BufferedOutput, so we buffer incoming records before emitting them to
Amazon Kinesis. You can find the detail about buffering mechanism [here](http://docs.fluentd.org/articles/buffer-plugin-overview).
Emitting records to Amazon Kinesis via network causes I/O Wait, so parallelizing
emitting with threads will improve throughput.

This option can be used to parallelize writes into the output(s)
designated by the output plugin. The default is 1.
Also you can use this option with *detach_process*.

### retries_on_putrecords

Integer, default is 3. When **order_events** is false, the plugin will put multiple
records to Amazon Kinesis in batches using PutRecords. A set of records in a batch
may fail for reasons documented in the Kinesis Service API Reference for PutRecords.
Failed records will be retried **retries_on_putrecords** times. If a record
fails all retries an error log will be emitted. 

### debug

Boolean. Enable if you need to debug Amazon Kinesis API call. Default is false.

## Configuration examples

Here are some configuration examles.
Assume that the JSON object below is coming to with tag 'your_tag'.

    {
      "name":"foo",
      "action":"bar"
    }

### Simply putting events to Amazon Kinesis with a partition key

In this example, simply a value 'foo' will be used as partition key,
then events will be sent to the stream specified in 'stream_name'.

    <match your_tag>
    type kinesis

    stream_name YOUR_STREAM_NAME

    aws_key_id YOUR_AWS_ACCESS_KEY
    aws_sec_key YOUR_SECRET_KEY

    region us-east-1

    partition_key name
    </match>

### Using partition_key_expr to add specific prefix to partition key

In this example, we add partition_key_expr to the example above.
This expression adds string 'some_prefix-' to partition key 'name',
then partition key finally will be 'some_prefix-foo'.

With specifying parition_key and parition_key_expr both,
the extracted value for partition key from JSON object will be
passed to your Ruby expression as a variable 'record'.

    <match your_tag>
    type kinesis

    stream_name YOUR_STREAM_NAME

    aws_key_id YOUR_AWS_ACCESS_KEY
    aws_sec_key YOUR_SECRET_KEY

    region us-east-1

    partition_key name
    partition_key_expr 'some_prefix-' + record
    </match>

### Using partition_key_expr to extract a value for partition key

In this example, we use only partition_key_expr to extract
a value for partition key. It will be 'bar'.

Specifying partition_key_expr without partition_key,
hash object that is converted from whole JSON object will be
passed to your Ruby expression as a variable 'record'.

    <match your_tag>
    type kinesis

    stream_name YOUR_STREAM_NAME

    aws_key_id YOUR_AWS_ACCESS_KEY
    aws_sec_key YOUR_SECRET_KEY

    region us-east-1

    partition_key_expr record['action']
    </match>

### Improving throughput to Amazon Kinesis

The achievable throughput to Amazon Kinesis is limited to single-threaded
PutRecord calls if **order_events** is set to true. By setting **order_events**
to false records will be sent to Amazon Kinesis in batches. When operating in
this mode the plugin can also be configured to execute in parallel.
The **detach_process** and **num_threads** configuration settings control
parallelism.

Please note that **order_events** option will be ignored if you choose to
use either **detach_process** or **num_threads**.

In case of the configuration below, you will spawn 2 processes.

    <match your_tag>
    type kinesis

    stream_name YOUR_STREAM_NAME
    region us-east-1

    detach_process 2

    </match>

You can also specify a number of threads to put.
The number of threads is bound to each individual processes.
So in this case, you will spawn 1 process which has 50 threads.

    <match your_tag>
    type kinesis

    stream_name YOUR_STREAM_NAME
    region us-east-1

    num_threads 50
    </match>

Both options can be used together, in the configuration below,
you will spawn 2 processes and 50 threads per each processes.

    <match your_tag>
    type kinesis

    stream_name YOUR_STREAM_NAME
    region us-east-1

    detach_process 2
    num_threads 50
    </match>

## Related Resources

* [Amazon Kinesis Developer Guide](http://docs.aws.amazon.com/kinesis/latest/dev/introduction.html)  


CHANGELOG.md
# CHANGELOG

## 0.3.3-rc

- **multi_records_per_put support**: Added support multiple rows of data in single put.
- **multi_records_separator support**: separator
- **endpoint support**: Added support local endpoint.

## 0.3.2

- **http_proxy support**: Added HTTP proxy support.

## 0.3.1

- **Fluentd v0.12 support**: We now support Fluentd v0.12.

## 0.3.0

- **Throughput improvement**: Added support for [PutRecords API](http://docs.aws.amazon.com/kinesis/latest/APIReference/API_PutRecords.html) by default.
- **Bug fix**: Removed redundant Base64 encoding of data for each Kinesis record emitted. Applications consuming these records will need to be updated accordingly.

## 0.2.0

- **Partition key randomization**: Added support for partition key randomization.
- **Throughput improvements**: Added support for spawning additional processes and threads to increase throughput to Amazon Kinesis.
- **AWS SDK for Ruby V2**: Added support for [AWS SDK for Ruby V2](https://github.com/aws/aws-sdk-core-ruby).

## 0.1.0

- Release on Github


test/plugin/test_out_kinesis.rb
# Copyright 2014 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"). You
# may not use this file except in compliance with the License. A copy of
# the License is located at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# or in the "license" file accompanying this file. This file is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
# ANY KIND, either express or implied. See the License for the specific
# language governing permissions and limitations under the License.

require 'helper'

class KinesisOutputTest < Test::Unit::TestCase
  def setup
    Fluent::Test.setup
  end

  CONFIG = %[
    aws_key_id test_key_id
    aws_sec_key test_sec_key
    stream_name test_stream
    region us-east-1
    partition_key test_partition_key
    endpoint test_endpoint
  ]

  def create_driver(conf = CONFIG, tag='test')
    Fluent::Test::BufferedOutputTestDriver
      .new(FluentPluginKinesis::OutputFilter, tag).configure(conf)
  end

  def create_mock_client
    client = mock(Object.new)
    mock(Aws::Kinesis::Client).new({}) { client }
    return client
  end

  def test_configure
    d = create_driver
    assert_equal 'test_key_id', d.instance.aws_key_id
    assert_equal 'test_sec_key', d.instance.aws_sec_key
    assert_equal 'test_stream', d.instance.stream_name
    assert_equal 'us-east-1', d.instance.region
    assert_equal 'test_partition_key', d.instance.partition_key
    assert_equal 'test_endpoint2', d.instance.endpoint
  end

  def test_configure_with_more_options

    conf = %[
      stream_name test_stream
      region us-east-1
      http_proxy http://proxy:3333/
      partition_key test_partition_key
      partition_key_expr record
      explicit_hash_key test_hash_key
      explicit_hash_key_expr record
      order_events true
    ]
    d = create_driver(conf)
    assert_equal 'test_stream', d.instance.stream_name
    assert_equal 'us-east-1', d.instance.region
    assert_equal 'http://proxy:3333/', d.instance.http_proxy
    assert_equal 'test_partition_key', d.instance.partition_key
    assert_equal 'Proc',
      d.instance.instance_variable_get(:@partition_key_proc).class.to_s
    assert_equal 'test_hash_key', d.instance.explicit_hash_key
    assert_equal 'Proc',
      d.instance.instance_variable_get(:@explicit_hash_key_proc).class.to_s
    assert_equal 'a',
      d.instance.instance_variable_get(:@partition_key_proc).call('a')
    assert_equal 'a',
      d.instance.instance_variable_get(:@explicit_hash_key_proc).call('a')
    assert_equal true, d.instance.order_events
    assert_equal nil, d.instance.instance_variable_get(:@sequence_number_for_ordering)
  end

  def test_mode_configuration

    conf = %[
      stream_name test_stream
      region us-east-1
      partition_key test_partition_key
    ]
    d = create_driver(conf)
    assert_equal(false, d.instance.order_events)
    assert_equal(false, d.instance.instance_variable_get(:@parallel_mode))

    conf = %[
      stream_name test_stream
      region us-east-1
      partition_key test_partition_key
      order_events true
    ]
    d = create_driver(conf)
    assert_equal(true, d.instance.order_events)
    assert_equal(false, d.instance.instance_variable_get(:@parallel_mode))

    conf = %[
      stream_name test_stream
      region us-east-1
      partition_key test_partition_key
      num_threads 1
    ]
    d = create_driver(conf)
    assert_equal(false, d.instance.order_events)
    assert_equal(false, d.instance.instance_variable_get(:@parallel_mode))

    conf = %[
      stream_name test_stream
      region us-east-1
      partition_key test_partition_key
      num_threads 2
    ]
    d = create_driver(conf)
    assert_equal(false, d.instance.order_events)
    assert_equal(false, d.instance.multi_records_per_put)
    assert_equal("\v", d.instance.multi_records_separator)
    assert_equal(true, d.instance.instance_variable_get(:@parallel_mode))

    conf = %[
      stream_name test_stream
      region us-east-1
      partition_key test_partition_key
      detach_process 1
    ]
    d = create_driver(conf)
    assert_equal(false, d.instance.order_events)
    assert_equal(true, d.instance.instance_variable_get(:@parallel_mode))

    conf = %[
      stream_name test_stream
      region us-east-1
      partition_key test_partition_key
      order_events true
      multi_records_per_put false
      multi_records_separator "#{11.chr}"
      detach_process 1
      num_threads 2
    ]
    d = create_driver(conf)
    assert_equal(false, d.instance.order_events)
    assert_equal(false, d.instance.multi_records_per_put)
    assert_equal("\"\v\"", d.instance.multi_records_separator)
    assert_equal(true, d.instance.instance_variable_get(:@parallel_mode))

  end

  def test_format

    d = create_driver

    data1 = {"test_partition_key"=>"key1","a"=>1,"time"=>"2011-01-02T13:14:15Z","tag"=>"test"}
    data2 = {"test_partition_key"=>"key2","a"=>2,"time"=>"2011-01-02T13:14:15Z","tag"=>"test"}

    time = Time.parse("2011-01-02 13:14:15 UTC").to_i
    d.emit(data1, time)
    d.emit(data2, time)

    d.expect_format({
      'data' => data1.to_json,
      'partition_key' => 'key1' }.to_msgpack
    )
    d.expect_format({
      'data' => data2.to_json,
      'partition_key' => 'key2' }.to_msgpack
    )

    client = create_mock_client
    client.describe_stream(stream_name: 'test_stream')
    client.put_records(
      stream_name: 'test_stream',
      records: [
        {
          data: data1.to_json,
          partition_key: 'key1'
        },
        {
          data: data2.to_json,
          partition_key: 'key2'
        }
      ]
    ) { {} }

    d.run
  end

  def test_order_events

    d = create_driver(CONFIG + "\norder_events true")

    data1 = {"test_partition_key"=>"key1","a"=>1,"time"=>"2011-01-02T13:14:15Z","tag"=>"test"}
    data2 = {"test_partition_key"=>"key2","a"=>2,"time"=>"2011-01-02T13:14:15Z","tag"=>"test"}

    time = Time.parse("2011-01-02 13:14:15 UTC").to_i
    d.emit(data1, time)
    d.emit(data2, time)

    d.expect_format({
      'data' => data1.to_json,
      'partition_key' => 'key1' }.to_msgpack
    )
    d.expect_format({
      'data' => data2.to_json,
      'partition_key' => 'key2' }.to_msgpack
    )

    client = create_mock_client
    client.describe_stream(stream_name: 'test_stream')
    client.put_record(
      data: data1.to_json,
      partition_key: 'key1',
      stream_name: 'test_stream'
    ) { {sequence_number: 1} }
    client.put_record(
      data: data2.to_json,
      partition_key: 'key2',
      sequence_number_for_ordering: 1,
      stream_name: 'test_stream'
    ) { {} }

    d.run
  end

  def test_format_at_lowlevel
    d = create_driver
    data = {"test_partition_key"=>"key1","a"=>1}
    assert_equal(
        MessagePack.pack({
            "data"              => data.to_json,
            "partition_key"     => "key1"
        }),
        d.instance.format('test','test',data)
    )
  end

  def test_format_at_lowlevel_with_more_options

    conf = %[
      stream_name test_stream
      region us-east-1
      partition_key test_partition_key
      partition_key_expr record
      explicit_hash_key test_hash_key
      explicit_hash_key_expr record
      order_events true
    ]

    d = create_driver(conf)
    data = {"test_partition_key"=>"key1","test_hash_key"=>"hash1","a"=>1}
    assert_equal(
        MessagePack.pack({
            "data"              => data.to_json,
            "partition_key"     => "key1",
            "explicit_hash_key" => "hash1"
        }),
        d.instance.format('test','test',data)
    )
  end

  def test_get_key
    d = create_driver
    assert_equal(
      "1",
      d.instance.send(:get_key, "partition_key", {"test_partition_key" => 1})
    )

    assert_equal(
      "abc",
      d.instance.send(:get_key, "partition_key", {"test_partition_key" => "abc"})
    )

    d = create_driver(%[
      random_partition_key true
      stream_name test_stream
      region us-east-1'
    ])

    assert_match(
      /\A[\da-f-]{36}\z/,
      d.instance.send(:get_key, 'foo', 'bar')
    )

    d = create_driver(%[
      random_partition_key true
      partition_key test_key
      stream_name test_stream
      region us-east-1'
    ])

    assert_match(
      /\A[\da-f-]{36}\z/,
      d.instance.send(
        :get_key,
        'partition_key',
        {"test_key" => 'key1'}
      )
    )

    d = create_driver(%[
      random_partition_key true
      partition_key test_key
      explicit_hash_key explicit_key
      stream_name test_stream
      region us-east-1'
    ])

    assert_match(
      /\A[\da-f-]{36}\z/,
      d.instance.send(
        :get_key,
        'partition_key',
        {"test_key" => 'key1', "explicit_key" => 'key2'}
      )
    )
  end

  def test_record_exceeds_max_size
    d = create_driver
    string = ''
    (1..1024).each{ string = string + '1' }
    assert_equal(
      false,
      d.instance.send(:record_exceeds_max_size?,string)
    )

    string = ''
    (1..(1024*50)).each{ string = string + '1' }
    assert_equal(
      false,
      d.instance.send(:record_exceeds_max_size?,string)
    )

    string = ''
    (1..(1024*51)).each{ string = string + '1' }
    assert_equal(
      true,
      d.instance.send(:record_exceeds_max_size?,string)
    )
  end

  def test_build_records_array_to_put
    d = create_driver

    data_list = []
    (0..500).each do |n|
      data_list.push({data: n.to_s})
    end
    result = d.instance.send(:build_records_array_to_put,data_list)
    assert_equal(2,result.length)
    assert_equal(500,result[0].length)
    assert_equal(1,result[1].length)

    data_list = []
    (0..1400).each do
      data_list.push({data: '1'})
    end
    result = d.instance.send(:build_records_array_to_put,data_list)
    assert_equal(3,result.length)
    assert_equal(500,result[0].length)
    assert_equal(500,result[1].length)
    assert_equal(401,result[2].length)

    data_list = []
    data_string = ''
    (0..(1024*30)).each do
      data_string = data_string + '1'
    end
    (0..500).each do
      data_list.push({data: data_string})
    end
    result = d.instance.send(:build_records_array_to_put,data_list)
    assert_equal(3,result.length)
    assert_equal(170,result[0].length)
    assert_operator(
      1024 * 1024 *5, :>,
      result[0].reduce(0){|sum,i| sum + i[:data].length}
    )
    assert_equal(170,result[1].length)
    assert_operator(
      1024 * 1024 *5, :>,
      result[1].reduce(0){|sum,i| sum + i[:data].length}
    )
    assert_equal(161,result[2].length)
    assert_operator(
      1024 * 1024 *5, :>,
      result[2].reduce(0){|sum,i| sum + i[:data].length}
    )
  end

  def test_build_records_by_partition_key
    d = create_driver

    data_list = []
    (1..500).each do |n|
      data_list.push({data: n.to_s, partition_key: (n % 3).to_s})
    end
    result = d.instance.send(:build_records_by_partition_key,data_list)
    partition_keys = result.map{|r| r[:partition_key]}.sort
    assert_equal(partition_keys, ["0", "1", "2"])
    array0_data_all = result.find{|r| r[:partition_key] == "0"}[:data]
    assert_instance_of(String, array0_data_all)
    assert_includes(array0_data_all, d.instance.multi_records_separator)
    array0_size = array0_data_all.split(d.instance.multi_records_separator).size
    assert_equal(166, array0_size)
    array1_size = result.find{|r| r[:partition_key] == "1"}[:data].split(d.instance.multi_records_separator).size
    assert_equal(167, array1_size)
    array2_size = result.find{|r| r[:partition_key] == "2"}[:data].split(d.instance.multi_records_separator).size
    assert_equal(167, array2_size)

    d = create_driver(%[
      random_partition_key true
      partition_key test_key
      stream_name test_stream
      region us-east-1'
    ])
    result = d.instance.send(:build_records_by_partition_key,data_list)
    assert_equal(1, result.size)
    assert_equal(data_list[0][:partition_key], result[0][:partition_key])
    data_array_size = result[0][:data].split(d.instance.multi_records_separator).size
    assert_equal(data_list.size, data_array_size)
  end

  def test_build_data_to_put
    d = create_driver
    assert_equal(
      {key: 1},
      d.instance.send(:build_data_to_put,{"key"=>1})
    )
  end

  def test_calculate_sleep_duration
    d = create_driver
    assert_operator(
        1, :>,
        d.instance.send(:calculate_sleep_duration,0)
    )
    assert_operator(
        2, :>,
        d.instance.send(:calculate_sleep_duration,1)
    )
    assert_operator(
        4, :>,
        d.instance.send(:calculate_sleep_duration,2)
    )
  end
end


lib/fluent/plugin/out_kinesis.rb
# Copyright 2014 Amazon.com, Inc. or its affiliates. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License"). You
# may not use this file except in compliance with the License. A copy of
# the License is located at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# or in the "license" file accompanying this file. This file is
# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
# ANY KIND, either express or implied. See the License for the specific
# language governing permissions and limitations under the License.

require 'aws-sdk-core'
require 'multi_json'
require 'logger'
require 'securerandom'
require 'fluent/plugin/version'

module FluentPluginKinesis
  class OutputFilter < Fluent::BufferedOutput

    include Fluent::DetachMultiProcessMixin
    include Fluent::SetTimeKeyMixin
    include Fluent::SetTagKeyMixin

    USER_AGENT_NAME = 'fluent-plugin-kinesis-output-filter'
    PROC_BASE_STR = 'proc {|record| %s }'
    PUT_RECORDS_MAX_COUNT = 500
    PUT_RECORD_MAX_DATA_SIZE = 1024 * 50
    PUT_RECORDS_MAX_DATA_SIZE = 1024 * 1024 * 5
    RECORDS_SEPARATOR = "\v"

    Fluent::Plugin.register_output('kinesis',self)

    config_set_default :include_time_key, true
    config_set_default :include_tag_key,  true

    config_param :aws_key_id,  :string, default: nil
    config_param :aws_sec_key, :string, default: nil
    # The 'region' parameter is optional because
    # it may be set as an environment variable.
    config_param :region,      :string, default: nil

    config_param :stream_name,            :string
    config_param :random_partition_key,   :bool,   default: false
    config_param :partition_key,          :string, default: nil
    config_param :partition_key_expr,     :string, default: nil
    config_param :explicit_hash_key,      :string, default: nil
    config_param :explicit_hash_key_expr, :string, default: nil
    config_param :order_events,           :bool,   default: false
    config_param :multi_records_per_put,  :bool,   default: false
    config_param :multi_records_separator, :string, default: RECORDS_SEPARATOR
    config_param :retries_on_putrecords,  :integer, default: 3

    config_param :debug, :bool, default: false

    config_param :http_proxy, :string, default: nil
    config_param :endpoint, :string, default: nil

    def configure(conf)
      super
      validate_params

      if @detach_process or (@num_threads > 1)
        @parallel_mode = true
        if @detach_process
          @use_detach_multi_process_mixin = true
        end
      else
        @parallel_mode = false
      end

      if @parallel_mode
        if @order_events
          log.warn 'You have set "order_events" to true, however this configuration will be ignored due to "detach_process" and/or "num_threads".'
        end
        @order_events = false
      end

      if @partition_key_expr
        partition_key_proc_str = sprintf(
          PROC_BASE_STR, @partition_key_expr
        )
        @partition_key_proc = eval(partition_key_proc_str)
      end

      if @explicit_hash_key_expr
        explicit_hash_key_proc_str = sprintf(
          PROC_BASE_STR, @explicit_hash_key_expr
        )
        @explicit_hash_key_proc = eval(explicit_hash_key_proc_str)
      end
    end

    def start
      detach_multi_process do
        super
        load_client
        check_connection_to_stream
      end
    end

    def format(tag, time, record)
      data = {
        data: record.to_json,
        partition_key: get_key(:partition_key,record)
      }

      if @explicit_hash_key or @explicit_hash_key_proc
        data[:explicit_hash_key] = get_key(:explicit_hash_key,record)
      end

      data.to_msgpack
    end

    def write(chunk)
      data_list = chunk.to_enum(:msgpack_each).find_all{|record|
        unless record_exceeds_max_size?(record['data'])
          true
        else
          log.error sprintf('Record exceeds the 50KB per-record size limit and will not be delivered: %s', record['data'])
          false
        end
      }.map{|record|
        build_data_to_put(record)
      }

      if @order_events
        put_record_for_order_events(data_list)
      else
        if @multi_records_per_put
          records = build_records_by_partition_key(data_list)
          put_records_with_retry(records)
        else
          records_array = build_records_array_to_put(data_list)
          records_array.each{|records|
            put_records_with_retry(records)
          }
        end
      end
    end

    private
    def validate_params
      unless @random_partition_key or @partition_key or @partition_key_expr
        raise Fluent::ConfigError, "'random_partition_key' or 'partition_key' or 'partition_key_expr' is required"
      end
    end

    def load_client

      user_agent_suffix = "#{USER_AGENT_NAME}/#{FluentPluginKinesis::VERSION}"

      options = {
        user_agent_suffix: user_agent_suffix
      }

      if @region
        options[:region] = @region
      end

      if @aws_key_id && @aws_sec_key
        options.update(
          access_key_id: @aws_key_id,
          secret_access_key: @aws_sec_key,
        )
      end

      if @debug
        options.update(
          logger: Logger.new(log.out),
          log_level: :debug
        )
        # XXX: Add the following options, if necessary
        # :http_wire_trace => true
      end

      if @http_proxy
        options[:http_proxy] = @http_proxy
      end

      if @endpoint
        options[:endpoint] = @endpoint
      end

      @client = Aws::Kinesis::Client.new(options)
    end

    def check_connection_to_stream
      @client.describe_stream(stream_name: @stream_name)
    end

    def get_key(name, record)
      if @random_partition_key
        SecureRandom.uuid
      else
        key = instance_variable_get("@#{name}")
        key_proc = instance_variable_get("@#{name}_proc")

        value = key ? record[key] : record

        if key_proc
          value = key_proc.call(value)
        end

        value.to_s
      end
    end

    def build_data_to_put(data)
      Hash[data.map{|k, v| [k.to_sym, v] }]
    end

    def put_record_for_order_events(data_list)
      sequence_number_for_ordering = nil
      data_list.each do |data_to_put|
        if sequence_number_for_ordering
          data_to_put.update(
            sequence_number_for_ordering: sequence_number_for_ordering
          )
        end
        data_to_put.update(
          stream_name: @stream_name
        )
        result = @client.put_record(data_to_put)
        sequence_number_for_ordering = result[:sequence_number]
      end
    end

    def build_records_array_to_put(data_list)
      records_array = []
      records = []
      records_payload_length = 0
      data_list.each{|data_to_put|
        payload = data_to_put[:data]
        if records.length >= PUT_RECORDS_MAX_COUNT or (records_payload_length + payload.length) >= PUT_RECORDS_MAX_DATA_SIZE
          records_array.push(records)
          records = []
          records_payload_length = 0
        end
        records.push(data_to_put)
        records_payload_length += payload.length
      }
      records_array.push(records)
      records_array
    end

    def build_records_by_partition_key(data_list)
      record_array = []

      if @random_partition_key
        record_list = data_list.map{|record| record[:data]}
        data_to_put = data_list[0].dup
        data_to_put[:data] = record_list.join(@multi_records_separator)
        record_array.push(data_to_put)
      else
        data_list.group_by{|d| d[:partition_key]}.each_value { |d_list|
          record_list = d_list.map{|record| record[:data]}
          data_to_put = d_list[0].dup
          data_to_put[:data] = record_list.join(@multi_records_separator)
          record_array.push(data_to_put)
        }
      end

      record_array
    end

    def put_records_with_retry(records,retry_count=0)
      response = @client.put_records(
        records: records,
        stream_name: @stream_name
      )

      if response[:failed_record_count] && response[:failed_record_count] > 0
        failed_records = []
        response[:records].each_with_index{|record,index|
          if record[:error_code]
            failed_records.push({body: records[index], error_code: record[:error_code]})
          end
        }

        if(retry_count < @retries_on_putrecords)
          sleep(calculate_sleep_duration(retry_count))
          retry_count += 1
          log.warn sprintf('Retrying to put records. Retry count: %d', retry_count)
          put_records_with_retry(
            failed_records.map{|record| record[:body]},
            retry_count
          )
        else
          failed_records.each{|record|
            log.error sprintf(
              'Could not put record, Error: %s, Record: %s',
              record[:error_code],
              record[:body].to_json
            )
          }
        end
      end
    end

    def calculate_sleep_duration(current_retry)
      Array.new(@retries_on_putrecords){|n| ((2 ** n) * scaling_factor)}[current_retry]
    end

    def scaling_factor
      0.5 + Kernel.rand * 0.1
    end

    def record_exceeds_max_size?(record_string)
      return record_string.length > PUT_RECORD_MAX_DATA_SIZE
    end
  end
end
